ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -348,6 +358,39 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 	return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
 }
 
+// checkBackoffLimitAndUpdateStatusIfNeeded determines if a RayJob is eligible for retry based on the configured backoff limit,
+// the job's success status, and its failure status. If eligible, sets the JobDeploymentStatus to Retrying.
+func (r *RayJobReconciler) checkBackoffLimitAndUpdateStatusIfNeeded(ctx context.Context, rayJob *rayv1.RayJob) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	failedCount := int32(0)
+	if rayJob.Status.Failed != nil {
+		failedCount = *rayJob.Status.Failed
+	}
+
+	succeededCount := int32(0)
+	if rayJob.Status.Succeeded != nil {
+		succeededCount = *rayJob.Status.Succeeded
+	}
+
+	if rayJob.Status.JobDeploymentStatus == rayv1.JobDeploymentStatusFailed {
+		failedCount++
+	}
+
+	if rayJob.Status.JobStatus == rayv1.JobStatusSucceeded && rayJob.Status.JobDeploymentStatus == rayv1.JobDeploymentStatusComplete {
+		succeededCount++
+	}
+
+	rayJob.Status.Failed = ptr.To[int32](failedCount)
+	rayJob.Status.Succeeded = ptr.To[int32](succeededCount)
+
+	if rayJob.Status.JobDeploymentStatus == rayv1.JobDeploymentStatusFailed && rayJob.Spec.BackoffLimit != nil && *rayJob.Status.Failed < *rayJob.Spec.BackoffLimit+1 {
+		logger.Info("RayJob is eligible for retry, setting JobDeploymentStatus to Retrying",
+			"backoffLimit", *rayJob.Spec.BackoffLimit, "succeeded", *rayJob.Status.Succeeded, "failed", rayJob.Status.Failed)
+		rayJob.Status.JobDeploymentStatus = rayv1.JobDeploymentStatusRetrying
```

Comment:
We should also check the reason for failure. If the reason is `DeadlineExceeded`, we should not retry.

Reply:
Good catch, I'll also have a follow-up PR to add tests for active deadline seconds + backoff limit

Reply:
Thinking about it more, it's kind of weird having to check `DeadlineExceeded` inside `checkBackoffLimitAndUpdateStatusIfNeeded`, it should be agnostic to the reason. 

What do you think about calling `checkBackoffLimitAndUpdateStatusIfNeeded` twice inside the `Running` switch case:
1) Right after calling checkK8sJobAndUpdateStatusIfNeeded
2) At the end of the Running switch case 

And we can exclude it after `checkActiveDeadlineAndUpdateStatusIfNeeded`. I think this will be less error prone since we're not checking retry for all future cases as well

Reply:
See the latest commit: https://github.com/ray-project/kuberay/pull/2192/commits/a7bc5cdf4f1f946b0ba2d53ecb578e991011dd28

Reply:
I may prefer to call `checkBackoffLimitAndUpdateStatusIfNeeded` once, but I am completely fine with merging this PR for now. I will open a follow up PR to update it a bit.